Horizontal morphing pdf for htt analyses

ajgilbert opened the pull request:
Hi, this is an initial version of a pdf that implements horizontal template morphing (A. Read method). The idea is to be able to build a single workspace for all mH, which can then be treated as any other POI. This is mostly just a convenience for the SM htt analysis, but is quite useful for the MSSM analyses where we want to be able to construct arbitrary multi-Higgs signal models as a function of mA and tanbeta (e.g. see slides 8-10 in https://indico.cern.ch/event/363307/session/20/contribution/31/material/slides/0.pdf)

Main features:
 - pdf builds a map: `map<double, FastVerticalInterpHistPdf2*>` from mass points to signal pdfs which handle the vertical morphing in the normal way
 - evaluate() checks for changes in the mass parameter, picks the appropriate pair of pdfs and morphs. Added methods to FastVerticalInterpHistPdf2 that allow us to check the status of the cache and access this directly
 - A known problem with the horizontal morphing is that if the histogram bins are wide relative to the width of the distribution the RMS is not interpolated smoothly (giving rise to bumpy limits). this can be fixed by morphing with finer binned templates and then rebinning - this is also supported in the pdf

comments & questions welcome

cheers,
Andrew

gpetruc commented on src/RooMorphingPdf.cc:
```diff
@@ -0,0 +1,589 @@
+  // Float_t const* dist1 = hist1.GetArray();
+  // Float_t const* dist2 = hist2.GetArray();
+
+  // AG: can re-use these too if nb1 and nb2 are fixed
```
I'd suggest to re-use the arrays rather than call new and delete within a  function that may be called multiple times within a minimization loop.
Even if `nb1` and `nb2` may not be fixed, but for sure they're bounded from above by a constant known at initialization time (the max number of bins in any template, I think)

nucleosynthesis commented:
What is the status of this? 
Can @ajgilbert implement the change suggested by Giovanni so we can merge?

ajgilbert commented:
Hi, sorry for the delay - I've implemented the suggested changes now. Although the original morphing was designed to support morphing between hists with different binning, this would add an unnecessary layer of complexity here, so I added a restriction that all inputs must have the same number of bins.  At some point, in a separate PR, I'll add some example code for making one of these pdfs, as right now it's probably only of use to people with the tautau datacard framework.

adavidzh commented:
A note regarding multidimensional morphing:
There is (a paper on this)[http://arxiv.org/abs/1410.7388] from ATLAS colleagues though the relevant `RooStarMomentMorph` and `RooMomentMorphND` seem to not yet be part of the latest ROOT distributions (certainly not in 5.34/32 and 6.04/00).
